Add parameters to mask analogy queries.

sebpuetz opened the pull request:
Addresses finalfusion/finalfusion-utils#9

The mask allows to include or exclude a, b or c in analogy queries like: a is to b like c is to d.

Seems like I don't have write access to the finalfusion repo, so this is coming from a fork.

danieldk commented on src/similarity.rs:
```diff
@@ -54,6 +54,15 @@ pub trait Analogy {
         limit: usize,
     ) -> Option<Vec<WordSimilarity>>;
+
+    fn analogy_masked(
```
The `analogy` method above could now provide a default implementation that calls `analogy_by` the `analogy` implementation in the `impl` block can then be removed.

danieldk commented on src/similarity.rs:
```diff
@@ -68,12 +77,31 @@ where
+            word2,
+            word3,
+            limit,
+            [false, false, false],
```
I am a bit confused. I thought being masked would mean that the word is not included in the result set. Then the 'traditional' analogy implementation should have `[true, true, true]` as the mask?

sebpuetz replied:
I initially had a failing CI build because I confused myself with this, too. 

I was thinking of masks as something like we use in tf models. true (or 1) for active steps and false (or 0) for inactive steps. Similarly here, false for those words that are inactive.

danieldk commented on src/similarity.rs:
```diff
@@ -120,7 +150,12 @@ where
+        let skip = [word1, word2, word3]
+            .iter()
+            .zip(mask.iter())
+            .filter(|(_, &include)| !include)
```
See above, I would expect masking to exclude the words (*masking* in the *conceal* sense). Perhaps mask is not the correct verb to use, maybe *exclude*?

sebpuetz commented:
Pushed the changes

danieldk commented:
> Pushed the changes

Did you see the suggestion about the name change (masked -> remove)?

Also, there is no `analogy_by` with default masking.

sebpuetz commented:
> Did you see the suggestion about the name change (masked -> remove)?

Missed that.

> Also, there is no `analogy_by` with default masking.

I thought you meant a default implementation for `Analogy::analogy`, which I added. 

Now I also added another method `AnalogyBy::analogy_by_masked` and a default impl for `AnalogyBy::analogy_by` with the default masking. Also, rustdoc was missing for the new method, which I added now.
